src/gluonts/dataset/arrow.py

```diff
@@ -0,0 +1,312 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+import re
+from pathlib import Path
+from typing import (
+    Union,
+    Dict,
+    List,
+    Any,
+    cast,
+    Optional,
+    Tuple,
+    Iterator,
+    Iterable,
+)
+
+import numpy as np
+import pandas as pd
+import pyarrow as pa
+from gluonts.dataset import util
+from gluonts.dataset.util import get_bounds_for_mp_data_loading
+from gluonts.gluonts_tqdm import tqdm
+
+from .common import Dataset, DataEntry, ProcessDataEntry, SourceContext
+
+_ARROW_NUMPY_SHAPE_SUFFIX = "._np_shape"
+
+
+class ArrowWriter:
+    """
+    Write records to an arrow table. The first record is used to infer the
+    schema. Lists are converted to numpy int or float arrays if possible.
+    Strings of the form 'yyyy-mm-dd ...' are converted to timestamps if
+    possible.
+
+    Note:
+    Numpy arrays with dimension > 1 are stored in two columns one with a
+    1d flat data an additional column called "<array_name>._np_shape" that
+    stores the array shape. One could also do this via pyarrow/pandas
+    extension types, but it seemed overly complicated for this use case.
+    """
+
+    def __init__(
+        self,
+        path: Union[str, Path],
+        chunk_size: int = 100,
+        int_dtype=np.int32,
+        float_dtype=np.float32,
+    ):
+        self.path: Path = Path(path)
+        self.chunk_size = chunk_size
+        self.int_dtype = int_dtype
+        self.float_dtype = float_dtype
+        self._cur_cols: Dict[str, List] = {}
+        self._cur_n = 0
+        self._numpy_shape_cols: Dict[str, List] = {}
+        self._cols_to_numpy_dtype: Dict[str, Any] = {}
+        self._timestamp_cols: List[str] = []
+        self._writer = None
+        self._schema = None
+        self._first_record = True
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
+        self.close()
+
+    def _infer_types(self, record):
+        for k, v in record.items():
+            self._cur_cols[k] = []
+            if isinstance(v, list):
+                try:
+                    tmp = np.asarray(v)
+                    dtype_kind = tmp.dtype.kind
+                except Exception:
+                    dtype_kind = None
+                if dtype_kind == "f":
+                    self._cols_to_numpy_dtype[k] = self.float_dtype
+                    v = tmp
+                if dtype_kind == "i":
+                    self._cols_to_numpy_dtype[k] = self.int_dtype
+                    v = tmp
+            if isinstance(v, np.ndarray) and v.ndim > 1:
+                self._numpy_shape_cols[k] = []
+            if isinstance(v, str) and re.match(r"^\s*\d{4}-\d{2}-\d{2}", v):
```

Comment:
This looks weird to me. Why only support years? And isn't this better left to the user to define the correct input types?